PIPELINE: It's the end of HoloPipelines as we know it

nielsboecker opened the pull request:
This PR reflects the gigantic #45 refactoring that I did to the HoloPipelines. I did a whole bunch of things and nothing is like it used to be. Hallelujah!

Some major points:
* complete re-architecture of all components
* implement several features that were kinda half-baked before
* fix numerous issues and problems
* implement dynamic pipeline loading
* do some basic error handling
* use `multiprocessing` instead of `subprocess`
* job state in-memory instead of POSTing to self
* job-specific data kept in job's own directory
* automatic garbage collection
* add logging to file per job and serve that at a new endpoint
* add type hints and docstrings
* add documentation
* dockerise application
* add Postman collection
* identify further problems and inconsistencies, for which I left TODOs or opened issues

> **Disclaimer 1**: There are still many TODOs in the code. The test coverage is now zero. This must be worked on within following issues and PRs.

> **Disclaimer 2**: I have mostly worked with the `bone_segmentation` pipeline and can confirm that it works end-too-end. The other ones @ansonwong9695 is still working at confirming nothing broke.

nielsboecker committed:
PIPELINE: Extract obj2gltf wrapper

nielsboecker committed:
PIPELINE: Extract http client

nielsboecker committed:
PIPELINE: Extract io_paths config

nielsboecker committed:
PIPELINE: Refactor io_paths config

nielsboecker committed:
PIPELINE: Fix imports

nielsboecker committed:
PIPELINE: Refactor io_paths config

nielsboecker committed:
PIPELINE: Extract receive_input task

nielsboecker committed:
PIPELINE: Extract dispatch_output task and accessor adapter

nielsboecker committed:
PIPELINE: Refactor accessor adapter

nielsboecker committed:
PIPELINE: Extract job status util

nielsboecker committed:
PIPELINE: Extract pipelines info util

nielsboecker committed:
PIPELINE: Delete unused compGetInput.py

nielsboecker committed:
PIPELINE: Refactor accessor adapter

nielsboecker committed:
PIPELINE: Merge compGetPipelineListInfo.py with similar component

nielsboecker committed:
PIPELINE: Extract numpy transformation service and refactor stuff

nielsboecker committed:
PIPELINE: Extract status garbage collector component

nielsboecker committed:
PIPELINE: Extract job status component

nielsboecker committed:
PIPELINE: Delete unused compJobCleanup.py

nielsboecker committed:
PIPELINE: Merge compJobStatus.py with similar component

nielsboecker committed:
PIPELINE: Extract format conversion service

nielsboecker committed:
PIPELINE: Delete unused compConvertDicom2nifti.py

nielsboecker committed:
PIPELINE: Refactor tasks module structure

nielsboecker committed:
PIPELINE: Extract split_to_separate_organs task

nielsboecker committed:
PIPELINE: Extract numpy to obj conversion to services

nielsboecker committed:
PIPELINE: Refactor obj2gltf wrapper

nielsboecker committed:
PIPELINE: Remove useless error message

nielsboecker committed:
PIPELINE: Extract nifty to numpy conversion to services

nielsboecker committed:
PIPELINES: Remove file deletion from conversion logic

nielsboecker committed:
PIPELINE: Extract nifty_transformation service

nielsboecker committed:
PIPELINE: Improve naming for dicom to numpy conversion

nielsboecker committed:
PIPELINE: Separate responsibilities in dicom to numpy conversion

nielsboecker committed:
PIPELINE: Extract dicom_file adapter

nielsboecker committed:
PIPELINE: Refactor

nielsboecker committed:
PIPELINE: Extract nifti_file adapter

nielsboecker committed:
PIPELINES: Fix typo nifty/nifti

nielsboecker committed:
PIPELINE: Separate responsibilities in dicom to nifti conversion

nielsboecker committed:
PIPELINE: Separate responsibilities of marching cubes and obj writing

nielsboecker committed:
PIPELINE: Extract glb_file adapter

nielsboecker committed:
PIPELINE: Remove deletion from glb_file adapter

nielsboecker committed:
PIPELINE: Remove unsupported draco conversion

nielsboecker committed:
PIPELINE: Refactor obj2gltf wrapper

nielsboecker committed:
PIPELINE: Make holostorage_accessor client instead of adapter

nielsboecker committed:
PIPELINE: Fix wrong type hint

nielsboecker committed:
PIPELINE: Refactor ct_lung_segmentation wrapper

This is an existing implementation, that we copied and made minor changes to. I changes the module structure to represent that, created a wrapper, and added a README to document the changes made to the original source code.

nielsboecker committed:
PIPELINE: Refactor lung_and_airway_segmentation wrapper

nielsboecker committed:
PIPELINES: Fix minor errors

nielsboecker committed:
PIPELINE: Only do logging config in entry point scripts

nielsboecker committed:
PIPELINE: Refactor tests directory

nielsboecker committed:
PIPELINE: Extract job_controller component

nielsboecker committed:
PIPELINE: Delete unused pipelineConfig.py

This is essentially a clone of pipelineController.py, even though with slightly improved code. I copied the code from pipelineConfig to pipelineController and deleted pipelineConfig.

nielsboecker committed:
PIPELINE: Refactor package structure

nielsboecker committed:
MISC: Update .flake8

- Ignore E203 rule as it conflicts with black formatting
- Add comments
- Copy to HoloPipelines subdirectory, as it wasn't recognised before

nielsboecker committed:
PIPELINE: Change dir name and disable yet another broken test

nielsboecker committed:
PIPELINE: Remove unused technical pipelines without use case

nielsboecker committed:
PIPELINE: Rename pipelines to reflect their semantics

nielsboecker committed:
PIPELINE: Remove unnecessary files

nielsboecker committed:
PIPELINE: Fix broken /pipelines endpoint

nielsboecker committed:
PIPELINE: Improve logging message style

nielsboecker committed:
PIPELINE: Use @as_json decorator and refactor server

nielsboecker committed:
PIPELINE: Refactor job_status and add basic error handling

nielsboecker committed:
UI: Update format of POST body to HoloPipelines

nielsboecker committed:
PIPELINE: Refactor and implement job_controller

nielsboecker committed:
PIPELINE: Add downsampling for all pipelines and add bugfix

This commit brings Udomkarn's changes from https://github.com/nbckr/HoloRepository-Core/pull/54. I copied it, instead of merging, because the file structure in the refactoring branch changed a lot.

nielsboecker committed:
PIPELINE: Create job-specific directories

nielsboecker committed:
PIPELINE: Fix bug in obj2gltf wrapper

nielsboecker committed:
PIPELINE: Move lung_and_airway_segmentation wrapper

nielsboecker committed:
PIPELINE: Fix typo

nielsboecker committed:
PIPELINE: Implement dynamic loading

nielsboecker committed:
PIPELINE: Implement proper status updates and logging to files

nielsboecker committed:
PIPELINE: Implement /status endpoint

nielsboecker committed:
UI: Prepare response type for /status endpoint

nielsboecker committed:
PIPELINE: Implement /log endpoint

nielsboecker committed:
PIPELINE: Implement garbage collection

nielsboecker committed:
PIPELINE: Implement POST to HoloStorageAccessor

nielsboecker committed:
PIPELINE: Handle errors in job controller instead of sudden death

nielsboecker committed:
PIPELINE: Extract Niftynet client

nielsboecker committed:
PIPELINE: Refactor receive_input task

nielsboecker committed:
PIPELINE: Refactor pipelines list

nielsboecker committed:
PIPELINE: Extract pipelines controller

nielsboecker committed:
PIPELINE: Improve job state (CREATED and QUEUED)

nielsboecker committed:
PIPELINE: Remove deprecated pipeline_runner

nielsboecker committed:
PIPELINE: Reactor various parts

nielsboecker committed:
PIPELINE: Delete unused io_paths component

nielsboecker committed:
PIPELINE: Refactor and improve split_to_separate_organs task

nielsboecker committed:
PIPELINE: Refactor various parts... again

nielsboecker committed:
PIPELINE: Refactor bone_segmentation pipeline

nielsboecker committed:
PIPELINE: Refactor lung_segmentation

Currently, when I test it with the normal-chest-mediastinal.zip input, I get this error in the PREPROCESSING state:

[ERROR] input and output rank must be > 0

nielsboecker committed:
PIPELINE: Refactor abdominal_organs_segmentation pipeline

Currently still having an issue when I test this locally on my PC, seems the model times out.

nielsboecker committed:
PIPELINE: Add Postman test collection

nielsboecker committed:
Merge branch 'dev' into PIPELINE/major-refactoring

nielsboecker committed:
PIPELINE: Add Dockerfile

nielsboecker committed:
PIPELINE: Clear out requirements.txt

nielsboecker committed:
PIPELINE: Remove pathlib dependency

This library has led to a lot of unnecessarily convoluted code, where actually Python is quite powerful and smart in accepting all kings of path-like objects. Just use strings in the future.

nielsboecker committed:
PIPELINE: Remove pathlib dependency

This library has led to a lot of unnecessarily convoluted code, where actually Python is quite powerful and smart in accepting all kinds of path-like objects. Just use strings in the future.

nielsboecker committed:
PIPELINE: Use environmnent variables for config

nielsboecker committed:
PIPELINE: Update Docker to take into account environment variables

nielsboecker committed:
PIPELINE: Fix unnecessary normalisation

Due to earlier design drawbacks, the input was normalised (a.k.a. "resampled") twice, which was confusing and unnecessary. The new design decouples responsibilities and thus allows to perform normalisation only when needed.

UdomkarnBoonyaprasert committed:
PIPELINE: fix bug in refactor scale

UdomkarnBoonyaprasert committed:
PIPELINE: delete comments

nielsboecker committed:
PIPELINE: Add module documentation

nielsboecker committed:
PIPELINE: Fix path name bug

nielsboecker committed:
Merge branch 'PIPELINE/major-refactoring' into PIPELINE/fix_scale_factor

nielsboecker committed:
Merge pull request #66 from nbckr/PIPELINE/fix_scale_factor

Pipeline: fix scale factor

UdomkarnBoonyaprasert committed:
PIPELINE: add normalisation to nifti

nielsboecker committed:
PIPELINE: Update README.

Add/update sections:
* Job-specific working areas
* How to add new pipelines
* API specification

boonwj commented on HoloPipelines/Dockerfile:
```diff
@@ -0,0 +1,25 @@
+ENTRYPOINT ["gunicorn"]
+
+# Note: Use only one gunicorn worker as we manually do multiprocessing
+# and having multiple workkers on gunicorn's end breaks the way we keep
```
workers

boonwj commented on HoloPipelines/README.md:
```diff
@@ -11,7 +11,7 @@ A typical pipeline will process the data fully automatic by utilising a NN model
 As each pipeline is independent, semi-automated or manual pipelines (which may even include their own front-ends or manual processing steps) could be added later.
 
-Upon receiving a DICOM image series, a job will be started and passed through the different stages of a pipeline. The status of each job can be queried through a distinct API. When finished, the result will be handed on to the HoloStorage Accessor.
+###### Upon receiving a DICOM image series, a job will be started and passed through the different stages of a pipeline. The status of each job can be queried through a distinct API. When finished, the result will be handed on to the HoloStorage Accessor.
```
Should this be a sub heading? Perhaps use quote (>) if it needs emphasis

nielsboecker replied:
I think this happened mistakenly, shouldn't be a subbheading.

boonwj commented on HoloPipelines/README.md:
```diff
@@ -34,12 +34,41 @@ The modules that form a pipeline can perform different tasks:
 
+## How to add new pipelines
+
+Adding a new pipeline is easy. The required steps vary depending on the specific use case, and on the type of pipeline. Currently, we have three types of pipelines, that differ in the way they perform automatic segmentation:
```
The ease of adding a new pipeline is hard to visualise with words. Perhaps a template could be provided somewhere to show the ease of implementation.

nielsboecker replied:
Would be nice, but may well be out of scope given that we have basically just a week left :)

boonwj commented on HoloPipelines/core/adapters/dicom_file.py:
```diff
@@ -0,0 +1,130 @@
+    return slices
+
+
+def read_dicom_pixels_as_np_ndarray(input_path):
```
missing type hint for variable

boonwj commented on HoloPipelines/core/adapters/nifti_file.py:
```diff
@@ -0,0 +1,94 @@
+    """
+    # TODO: Why copy??
+    image = image_data
+    # TODO: does the Nifti1Image have a .shape field? according to documentary, no?
```
documentation*

boonwj commented on HoloPipelines/core/adapters/obj_file.py:
```diff
@@ -0,0 +1,19 @@
+This module contains functionality related to writing a mesh to disk as an OBJ file.
+"""
+
+def write_mesh_as_obj(verts, faces, norm, output_obj_path):
```
missing type hints

boonwj commented on HoloPipelines/core/clients/holostorage_accessor.py:
```diff
@@ -0,0 +1,58 @@
+# directories and less stuff has to be handed from component to component.
+
+holostorage_baseurl = f"{HOLOSTORAGE_ACCESSOR_HOST}:{HOLOSTORAGE_ACCESSOR_PORT}"
+holograms_endpoint = f"{holostorage_baseurl}/api/v1/holograms"
```
api version should be configurable

boonwj commented on HoloPipelines/core/tasks/abdominal_organs_segmentation/split_to_separate_organs.py:
```diff
@@ -0,0 +1,65 @@
+
+def get_organ_name(hu_value: int):
+    """
+    Maps the "HU valuee" to the according organ name. Note that this is just an
```
value*

boonwj commented on HoloPipelines/core/tasks/abdominal_organs_segmentation/split_to_separate_organs.py:
```diff
@@ -0,0 +1,65 @@
+of splitting the combined result into multiple sub-results.
+"""
+
+# TODO: Depending on how we deciide for this pipeline, this component may get removed.
```
decide*

nielsboecker committed:
Update README.md

boonwj reviewed (commented):
Just a quick eyeball through for generic typos and inconsistencies. 

Overall, code is clean and pleasant to read, but I haven't grasped the way everything works yet. I will continue looking deeper through the changes.

UdomkarnBoonyaprasert committed:
PIPELINE: delete commented out code

UdomkarnBoonyaprasert committed:
Merge branch 'PIPELINE/major-refactoring' into PIPELINE/fix_nifti_normalise

boonwj commented on HoloPipelines/server.py:
```diff
@@ -0,0 +1,74 @@
+from jobs.jobs_io import read_log_file_for_job
+
+log_format = "%(asctime)s | %(name)s | %(levelname)s | %(message)s'"
+coloredlogs.install(level=logging.DEBUG, fmt=log_format)
```
Log level hard coded to DEBUG might not be ideal when switched to production mode. Could perhaps be based on the configured FLASK_ENV configuration.
```suggestion
log_level = logging.DEBUG if FLASK_ENV == "development" else logging.INFO
coloredlogs.install(level=log_level, fmt=log_format)
```

nielsboecker committed:
PIPELINE: Fix config parsing bug

Boolean and numerical values were parsed as strings, d'oh

nielsboecker committed:
PIPELINE: Remove redundant value casts of config variables

nielsboecker committed:
PIPELINE: Replace in-memory global state with file-based solution

Keeping state in-memory is hard with concurrency. At the moment we
have multiple worker processes in job_controller, and potentially
another level of concurrenty when gunicorn runs Flask. Also, this
is more in line with the rest of the state management.

nielsboecker committed:
PIPELINE: Update README.md to reflect new state-keeping method

nielsboecker committed:
PIPELINE: Add support for reading logs from finished jobs

boonwj committed:
Merge pull request #71 from nbckr/PIPELINE/fix-config-bug

PIPELINE: Fix config bug

UdomkarnBoonyaprasert committed:
PIPELINE: delete unused code

nielsboecker committed:
PIPELINE: Handle some edge cases for jobs

UdomkarnBoonyaprasert committed:
PIPELINE: fix conflicts

nielsboecker committed:
PIPELINE: Extract setup methods from __main__

ansonwong9695 committed:
Add conversion from numpy to obj and obj to glb

ansonwong9695 committed:
Add conversion from numpy to obj and obj to glb

nielsboecker committed:
Update abdominal_organs_segmentation.py

* cast to int not necessary anymore with new changes on refactoring branch
* moved comment

nielsboecker committed:
Merge pull request #74 from nbckr/PIPELINE/complete-abdominal-pipeline

Pipeline: complete abdominal pipeline

nielsboecker committed:
PIPELINE: Fix typo and implement PR suggestions

nielsboecker committed:
PIPELINE: Add return type hints

UdomkarnBoonyaprasert committed:
PIPELINE: add testing for nifti_file
testing still local at this point, will update in next commit

UdomkarnBoonyaprasert committed:
PIPELINE: update test file
update sample nifti file with one from blob
and add test_setup

UdomkarnBoonyaprasert committed:
PIPELINE: minor cleanup
removed unused code

boonwj commented on HoloPipelines/core/pipelines/pipelines_controller.py:
```diff
@@ -6,26 +6,29 @@
-    Loads a pipeline module dynamically, which can then be invoked by its run function.
+    Loads and returns a pipeline module dynamically, which can then be invoked by its
+    run
+    function.
```
This does not look intended.

boonwj commented on HoloPipelines/core/wrappers/obj2gltf.py:
```diff
@@ -2,14 +2,14 @@
 
 
-def call_obj2gltf(obj_input_path: str, glb_output_path: str):
+def call_obj2gltf(obj_input_filepath: str, glb_output_file_path: str) -> None:
```
filepath or file_path?

boonwj replied:
stick to `file_path`, it seems to be used more commonly in the rest of the codebase

nielsboecker committed:
Merge branch 'PIPELINE/major-refactoring' into PIPELINE/gunicorn-entrypoint

nielsboecker committed:
Merge branch 'PIPELINE/major-refactoring' into PIPELINE/disk-state-improvements

nielsboecker committed:
PIPELINE: Follow up feedback from PR #67

UdomkarnBoonyaprasert committed:
PIPELINE: add test cases for nifti normalise

UdomkarnBoonyaprasert committed:
PIPELINE: fix condition for cropping

boonwj commented:
`holopipelines-core_1                             | /usr/local/lib/python3.7/site-packages/dotenv/main.py:52: UserWarning: File doesn't exist 
holopipelines-core_1                             |   warnings.warn("File doesn't exist {}".format(self.dotenv_path))  # type: ignore
`

This warning will be thrown in docker-compose. There are no .env files in docker compose as the .env run  We might want to document this in the README somewhere.

nielsboecker committed:
UI/BE: Use real HoloPipelines /pipelines endpoint

nielsboecker committed:
UI/BE: Change redirect HTTP code to 307

Code 308 was meant to represent a permanent redirect. While
semantically accurate, this lead to some difficulties in
development. Changed to 307, as it is non-cacheable.

nielsboecker committed:
PIPELINE: Allow CORS

boonwj commented:
```
holopipelines-core_1                             | [2019-08-26 11:25:14 +0000] [1] [CRITICAL] WORKER TIMEOUT (pid:9)
holopipelines-core_1                             | [2019-08-26 11:25:14 +0000] [9] [INFO] Worker exiting (pid: 9)
holopipelines-core_1                             | [2019-08-26 11:25:14 +0000] [17] [INFO] Booting worker with pid: 17
holopipelines-core_1                             | /usr/local/lib/python3.7/site-packages/dotenv/main.py:52: UserWarning: File doesn't exist 
holopipelines-core_1                             |   warnings.warn("File doesn't exist {}".format(self.dotenv_path))  # type: ignore
holopipelines-core_1                             | WARNING:root:Started process pool with 4 worker processes
holopipelines-models__dense_vnet_abdominal_ct_1  | [2019-08-26 11:25:14 +0000] [1] [CRITICAL] WORKER TIMEOUT (pid:8)
holopipelines-models__dense_vnet_abdominal_ct_1  | [2019-08-26 11:25:14 +0000] [1] [CRITICAL] WORKER TIMEOUT (pid:9)
holopipelines-models__dense_vnet_abdominal_ct_1  | [2019-08-26 11:25:14 +0000] [1] [CRITICAL] WORKER TIMEOUT (pid:10)
holopipelines-models__dense_vnet_abdominal_ct_1  | [2019-08-26 11:25:14 +0000] [1] [CRITICAL] WORKER TIMEOUT (pid:12)
holopipelines-models__dense_vnet_abdominal_ct_1  | [2019-08-26 11:25:14 +0000] [8] [INFO] Worker exiting (pid: 8)
holopipelines-models__dense_vnet_abdominal_ct_1  | [2019-08-26 11:25:14 +0000] [9] [INFO] Worker exiting (pid: 9)
holopipelines-models__dense_vnet_abdominal_ct_1  | [2019-08-26 11:25:14 +0000] [12] [INFO] Worker exiting (pid: 12)
holopipelines-models__dense_vnet_abdominal_ct_1  | [2019-08-26 11:25:14 +0000] [10] [INFO] Worker exiting (pid: 10)
holopipelines-models__dense_vnet_abdominal_ct_1  | [2019-08-26 11:25:14 +0000] [16] [INFO] Booting worker with pid: 16
holopipelines-models__dense_vnet_abdominal_ct_1  | [2019-08-26 11:25:14 +0000] [17] [INFO] Booting worker with pid: 17
holopipelines-models__dense_vnet_abdominal_ct_1  | [2019-08-26 11:25:14 +0000] [18] [INFO] Booting worker with pid: 18
holopipelines-models__dense_vnet_abdominal_ct_1  | [2019-08-26 11:25:14 +0000] [20] [INFO] Booting worker with pid: 20
holopipelines-core_1                             | [2019-08-26 11:53:36 +0000] [1] [CRITICAL] WORKER TIMEOUT (pid:17)
holopipelines-core_1                             | [2019-08-26 11:53:36 +0000] [17] [INFO] Worker exiting (pid: 17)
holopipelines-core_1                             | [2019-08-26 11:53:36 +0000] [25] [INFO] Booting worker with pid: 25
holopipelines-core_1                             | /usr/local/lib/python3.7/site-packages/dotenv/main.py:52: UserWarning: File doesn't exist 
holopipelines-core_1                             |   warnings.warn("File doesn't exist {}".format(self.dotenv_path))  # type: ignore
holopipelines-core_1                             | WARNING:root:Started process pool with 4 worker processes
holopipelines-models__dense_vnet_abdominal_ct_1  | [2019-08-26 11:53:36 +0000] [1] [CRITICAL] WORKER TIMEOUT (pid:16)
holopipelines-models__dense_vnet_abdominal_ct_1  | [2019-08-26 11:53:36 +0000] [1] [CRITICAL] WORKER TIMEOUT (pid:17)
holopipelines-models__dense_vnet_abdominal_ct_1  | [2019-08-26 11:53:36 +0000] [1] [CRITICAL] WORKER TIMEOUT (pid:18)
holopipelines-models__dense_vnet_abdominal_ct_1  | [2019-08-26 11:53:36 +0000] [1] [CRITICAL] WORKER TIMEOUT (pid:20)
holopipelines-models__dense_vnet_abdominal_ct_1  | [2019-08-26 11:53:36 +0000] [16] [INFO] Worker exiting (pid: 16)
holopipelines-models__dense_vnet_abdominal_ct_1  | [2019-08-26 11:53:36 +0000] [18] [INFO] Worker exiting (pid: 18)
holopipelines-models__dense_vnet_abdominal_ct_1  | [2019-08-26 11:53:36 +0000] [20] [INFO] Worker exiting (pid: 20)
holopipelines-models__dense_vnet_abdominal_ct_1  | [2019-08-26 11:53:36 +0000] [17] [INFO] Worker exiting (pid: 17)
holopipelines-models__dense_vnet_abdominal_ct_1  | [2019-08-26 11:53:36 +0000] [24] [INFO] Booting worker with pid: 24
holopipelines-models__dense_vnet_abdominal_ct_1  | [2019-08-26 11:53:36 +0000] [25] [INFO] Booting worker with pid: 25
holopipelines-models__dense_vnet_abdominal_ct_1  | [2019-08-26 11:53:36 +0000] [27] [INFO] Booting worker with pid: 27
holopipelines-models__dense_vnet_abdominal_ct_1  | [2019-08-26 11:53:36 +0000] [29] [INFO] Booting worker with pid: 29
```

Noticed that pipelines and models subprocesses are timing out with exceptions. This happened when I left the images up during testing. Any idea what could be the case? 

I am not sure if this actually causes any issues with the actual workflow.

nielsboecker committed:
UI: Adjust mapping to pipelines format

nbckr commented:
I didn't really encounter this as I mostly developed local (when I start the Docker daemon I can't do anything else anymore due to my PC being an old man).

Looks scary though. Could it be related to [this](https://github.com/benoitc/gunicorn/issues/1801#issuecomment-490182440) or [this](https://github.com/benoitc/gunicorn/issues/1801#issuecomment-489680514)?

boonwj committed:
Merge pull request #72 from nbckr/PIPELINE/disk-state-improvements

PIPELINE: disk state improvements

Merged to be reviewed in a single refactor PR.

boonwj committed:
Merge pull request #73 from nbckr/PIPELINE/gunicorn-entrypoint

PIPELINE: gunicorn entrypoint

Merging this to be evaluated as a whole in #67

boonwj committed:
Merge branch 'dev' into PIPELINE/major-refactoring

Bring in fixes and refactors introduced in dev

nielsboecker committed:
UI/BE: Update HoloPipelines /jobs endpoint

nielsboecker committed:
PIPELINE: Fix bug from merging #72's branch

UdomkarnBoonyaprasert committed:
PIPELINE: update variable names

UdomkarnBoonyaprasert committed:
PIPELINE: add testing for np_iamge_manipulation
- add testing
- bug fix in condional crop

boonwj committed:
PIPELINE: Fix port in NN model

boonwj committed:
PIPELINE: Network link for pipeline and nn model

The 2 components are unable to talk to one another without this

nielsboecker committed:
PIPELINE: Refactor np_image_manipulation_service test

nielsboecker committed:
Merge pull request #77 from nbckr/PIPELINE/fix_cropping

Pipeline: Update NumPy array cropping condition

boonwj committed:
PIPELINE: Add link between pipeline and accessor

boonwj committed:
PIPELINE: Add obj2gltf dependency installation

boonwj committed:
Merge branch 'PIPELINE/major-refactoring' of https://github.com/nbckr/HoloRepository-Core into PIPELINE/major-refactoring

nielsboecker committed:
PIPELINE: Add explicit log for load_env() [#67]

boonwj commented:
```
holopipelines-core_1                             | 2019-08-27 13:52:16 | root | INFO  | Success! Created hologram: {'hid': 'ef9ce292-7312-4d02-9243-46486105bc79', 'title': 'asdf', 'contentType': 'model/gltf-binary', 'fileSizeInKb': 2011, 'creationDate': '2019-08-27T13:52:04Z', 'creationMode': 'GENERATE_FROM_IMAGING_STUDY', 'creationDescription': 'Generated by HoloPipelines with the abdominal_organs_segmentation pipeline', 'aid': 'a102', 'pid': 'p100'}
holopipelines-core_1                             | 2019-08-27 13:52:16 | 450c08a31d1b4e3b | INFO  | [450c08a31d1b4e3b] Finished state DISPATCHING_OUTPUT in 11.991149663925171 seconds
holopipelines-core_1                             | 2019-08-27 13:52:16 | 450c08a31d1b4e3b | INFO  | [450c08a31d1b4e3b] Entering state => FINISHED
holopipelines-core_1                             | Exception in thread Thread-3:
holopipelines-core_1                             | Traceback (most recent call last):
holopipelines-core_1                             |   File "/usr/local/lib/python3.7/threading.py", line 926, in _bootstrap_inner
holopipelines-core_1                             |     self.run()
holopipelines-core_1                             |   File "/usr/local/lib/python3.7/threading.py", line 870, in run
holopipelines-core_1                             |     self._target(*self._args, **self._kwargs)
holopipelines-core_1                             |   File "/usr/local/lib/python3.7/multiprocessing/pool.py", line 486, in _handle_results
holopipelines-core_1                             |     cache[job]._set(i, obj)
holopipelines-core_1                             |   File "/usr/local/lib/python3.7/multiprocessing/pool.py", line 662, in _set
holopipelines-core_1                             |     self._callback(self._value)
holopipelines-core_1                             | TypeError: job_success_callback() takes 0 positional arguments but 1 was given
holopipelines-core_1                             | 
holopipelines-core_1                             | 2019-08-27 13:52:29 | root | INFO  | Global state | 1 jobs active: ['450c08a31d1b4e3b']
holopipelines-core_1                             | 2019-08-27 13:52:29 | root | INFO  | Garbage collection | Removing job '450c08a31d1b4e3b' (success=True)
holopipelines-core_1                             | 2019-08-27 13:52:29 | root | INFO  | Removing temporary files for job '450c08a31d1b4e3b
holopipelines-core_1                             | 2019-08-27 13:52:29 | root | INFO  | Moving './__jobs__/450c08a31d1b4e3b' to './__finished_jobs__/450c08a31d1b4e3b'
```
There's this one error that occurs at the end of the abdominal pipeline. Will test it with the 2 other pipelines too.

nbckr commented:
Hey @boonwj that is because the `abdominal_pipeline` is the only one to (unreasonably, because unused) return something at the end of its run. This will confuse the success callback handler in `jobs_controller`. I analysed this issue when you mentioned it yesterday; the only thing confusing me is that the state in the `major-refactoring` branch ([here](https://github.com/nbckr/HoloRepository-Core/blob/PIPELINE/major-refactoring/HoloPipelines/core/pipelines/abdominal_organs_segmentation.py)) already has the return statement removed, so this shouldn't occur if you built the container from that state?

boonwj commented:
Alright, me remove all my current images and rebuild again

ansonwong9695 committed:
Merge branch 'PIPELINE/major-refactoring' into PIPELINE/fix_nifti_normalise

boonwj commented on HoloPipelines/config.py:
```diff
@@ -0,0 +1,65 @@
+# Load data from .env file (will not override actual env variables if set)
+# Note that the .env file is .dockerignored, so in production it is expected
+# to set these variables throug the deployment workflow.
+logging.info("Attempting to load environment variables from .env file, if it exists")
```
Log level settings are only set after importing the config file. So since default log levels do not show info, this log message will not be seen.

boonwj replied:
I think a more elegant way will be to use a try, catch and then raise a warning when dotenv files are not found.

nielsboecker committed:
PIPELINE: Remove load_env() warning if not found [#67]

nielsboecker committed:
Merge branch 'PIPELINE/major-refactoring' into PIPELINE/fix_nifti_normalise

nielsboecker committed:
PIPELINE: Refactor test_nifti_file

nielsboecker committed:
PIPELINE: Refactor test_nifti_file and extract http client

nielsboecker committed:
Merge pull request #69 from nbckr/PIPELINE/fix_nifti_normalise

Pipeline: fix nifti normalisation

nielsboecker committed:
PIPELINE: Update lung_segmentation

boonwj committed:
Merge pull request #84 from nbckr/PIPELINE/fix-lung-segmentation

PIPELINE: Fix lung segmentation

boonwj committed:
PIPELINE: Add fix for callback function

Add argument that is expected in callback functions.

boonwj committed:
PIPELINE: use correct type hint for Any

Previous usage of any is a built-in function not a type

boonwj committed:
Merge pull request #86 from nbckr/PIPELINE/callback_fix

Pipeline/callback fix

boonwj committed:
Merge branch 'dev' into PIPELINE/major-refactoring

boonwj reviewed (approved):
Tested that basic workflows work in my local docker-compose setup. Will merge to allow for kubes deployment.
